### ISSUE TITLE
Feature request: Offline cell tower positioning

### ISSUE THREAD

**Issue by markpayneatwork** [mendhak/gpslogger#284]
Hi,

I am a big fan of the cell tower location feature in GPSlogger. However, when I travel with my mobile in Europe, I generally turn the data  feature off to avoid the horrendous roaming costs. This unfortunattely makes it impossible to use GPSlogger's cell location feature. I would therefore like to propose a feature that logs the cell-tower ID but doesn't convert it to a lat-long. The post-processing of the cell tower ids to gett long-lat coordinates could then be instigated  at a later time by the user  once they have an affordable data connection..

Cheers,

Mark

**Comment by E3V3A** [comment]
+1

In addition it would be a great feature to be able to add your own tags by dragging tag drop to any location of interest. It is not clear from the documentation if this is possible.

**Comment by runfalk** [comment]
I would really like this feature. My suggestion would be to optionally log both WiFi networks with signal strength and cell towers to a separate XML (or similar) file. That way one could use a service such as [MLS](https://location.services.mozilla.com/) to fill in gaps in the GPS retroactively, or contribute to the database.

**Comment by lenucksi** [comment]
Maybe the UnifiedNlp providers with the Mozilla backend could help you here, they would then not require any changes to GpsLogger

**Comment by runfalk** [comment]
Great suggestion! It would help for my use case, though I would prefer to have the data for later analysis and comparison between providers. It also seems to require me to run a custom ROM or root my phone, which I would rather not have to do.

**Comment by dragetd** [comment]
Yes, GPSLogger is using the positioning-API of Google. Scanning for Cell-Tower and keeping your own database requires some storage and significant more code to what GPSLogger is doing right now.

But as mentioned before, there is a solution! The @lenucksi linked the UnifiledNLP packages… these are an alternative backend to provide the location. Since this is a backend, *any* app can use it!

But you ROM must support UnifiedNLP. For example microG + Linage comes with it. Or there is also an XPosed module to add it to other ROMs.

There are module for location via wifi or cell tower, including offline-database support. But this is all out of scope of GPSLogger.

**Comment by alex9099** [comment]
Is this on the list of "things to be implemented"? would be nice to see this implemented

**Comment by runfalk** [comment]
I don't think there is such a list. As far as I understand the project is not in active development. From the README:

> Note: Development has paused, and the app is not available from the Play Store, see this note. The app is now on F-Droid.

**Comment by mendhak** [comment]
Yes sorry about this - these days I'm just doing some minor housekeeping here and there on the app.  I've updated F-Droid a few times and updated the SDK recently to accommodate for increasing location restrictions on newer Android versions. 

I could have a read through some documentation anyway if someone could point me at what API calls can be used to get Cell Tower Info, I did a verrryyy brief search and there seem to be various deprecated APIs?

Also if there are any sample files, or examples of expected output, that I could look at, that would be useful in understanding this.  And how do you process those files, to convert it to a set of locations?  Custom scripts?

**Comment by khimaros** [comment]
does the FDroid release use location data from UnifiedNlp or is that disabled due to the FCM bits being removed?

**Comment by khimaros** [comment]
filed a separate issue for UnifiedNlp support in the FDroid release here: https://github.com/mendhak/gpslogger/issues/1045
